src/core/ext/xds/xds_http_filters.h

```diff
@@ -0,0 +1,111 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GRPC_CORE_EXT_XDS_XDS_HTTP_FILTERS_H
+#define GRPC_CORE_EXT_XDS_XDS_HTTP_FILTERS_H
+
+#include <grpc/support/port_platform.h>
+
+#include <memory>
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "google/protobuf/any.upb.h"
+#include "upb/def.h"
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/channel/channel_stack.h"
+#include "src/core/lib/json/json.h"
+
+namespace grpc_core {
+
+extern const char* kXdsHttpRouterFilterConfigName;
+
+class XdsHttpFilterImpl {
+ public:
+  struct FilterConfig {
+    absl::string_view config_proto_type_name;
+    Json config;
+
+    bool operator==(const FilterConfig& other) const {
+      return config_proto_type_name == other.config_proto_type_name &&
+             config == other.config;
+    }
+    std::string ToString() const {
+      return absl::StrCat("{config_proto_type_name=", config_proto_type_name,
+                          " config=", config.Dump(), "}");
+    }
+  };
+
+  virtual ~XdsHttpFilterImpl() = default;
+
+  // Filter config protobuf type name.
+  virtual absl::string_view config_proto_type_name() const = 0;
+
+  // Loads the proto message into the upb symtab.
+  virtual void PopulateSymtab(upb_symtab* symtab) const = 0;
+
+  // Generates a Config from the xDS filter config proto.
+  // Used for the top-level config in the HCM HTTP filter list.
+  virtual absl::StatusOr<FilterConfig> GenerateFilterConfig(
+      upb_strview serialized_xds_config, upb_arena* arena) const = 0;
+
+  // Generates a Config from the xDS filter config proto.
+  // Used for the typed_per_filter_config override in VirtualHost and Route.
+  virtual absl::StatusOr<FilterConfig> GenerateFilterConfigOverride(
+      upb_strview serialized_xds_config, upb_arena* arena) const = 0;
+
+  // C-core channel filter implementation.
+  virtual const grpc_channel_filter* channel_filter() const = 0;
+
+  // Modifies channel args that may affect service config parsing (not
+  // visible to the channel as a whole).
+  // Takes ownership of args.  Caller takes ownership of return value.
+  virtual grpc_channel_args* ModifyChannelArgs(grpc_channel_args* args) const {
+    return args;
+  }
+
+  // Function to convert the Configs into a JSON string to be added to the
+  // per-method part of the service config.
+  // The hcm_filter_config comes from the HttpConnectionManager config.
+  // The filter_config_override comes from the first of the VirtualHost,
+  // Route, or ClusterWeight entries that it is found in, or null if
+  // there is no override in any of those locations.
+  virtual absl::StatusOr<std::string> GenerateServiceConfig(
+      const FilterConfig& hcm_filter_config,
+      const FilterConfig* filter_config_override) const = 0;
```

Comment:
What will be the expected behavior for merging those two filter configs?

Reply:
As I mentioned elsewhere, it's up to the filter implementation how to do this.  This is not something we can determine for it.

Reply:
Envoy doesn't proactively merge the configs that way that gRPC will do.  Instead, it provides an API called `resolveMostSpecificPerFilterConfig()` for individual filters to explicitly request their override config when they want it, which ultimately calls the following code:

https://github.com/envoyproxy/envoy/blob/17e815122ff53d0ac6cb2d64cdbf1bfc547bb7e8/source/common/http/utility.cc#L915

There's apparently also a `mostSpecificPerFilterConfigTyped()` that does basically the same thing as `resolveMostSpecificPerFilterConfig()`, although the code is a lot harder to follow.

Reply:
Thanks for digging into their code base. Even with code search, I can't be certain if there is no "merge" logic or I missed it.